### ISSUE TITLE
【论文复现】损失函数的更改

### ISSUE THREAD

**Issue by scott0o0** [PaddlePaddle/Paddle#26681]
- 标题：动态图更改损失函数
- 版本、环境信息：
    1）PaddlePaddle版本：1.8
    3）GPU：2080ti , cuda10.0
    4）系统环境：windows , python3.7
- 训练信息
    1）单机，单卡

- 复现信息：想更改损失函数进行训练。
- 问题描述：想对yolov3计算的损失函数loss = fluid.layers.yolov3_loss()进行补充，增加一项损失与原先的损失之和作为总损失函数total_loss = loss + loss_add。其中loss_add中有一项是需要计算两个矩阵的欧式距离||X-Y||，如图1：
![image](https://user-images.githubusercontent.com/49825779/90511958-e4000180-e18f-11ea-91ab-8853041df1f8.png)
可我无法做到保留loss_add变量中的梯度特征（stop_gradient=False ），由于是利用动态图编写的方式，不知道如何自定义损失函数。

注：总损失函数total_loss = loss + loss_add中，loss用的是飞桨已有的损失函数fluid.layers.yolov3_loss()，loss_add是我用numpy写的（写完转换成的tensor变量），输入数据都是同一个主干网络中的不同层级的输出（tensor变量）如图2：
![image](https://user-images.githubusercontent.com/49825779/90401260-45609b80-e0d0-11ea-8e6a-07ecd4101eea.png)

**Comment by zhiqiu** [comment]
hi，我理解你需要将loss_add的逻辑用paddle的接口写，这样在自动求导的时候才能将梯度传过去。否则，paddle是不能将梯度传到numpy数组的。

**Comment by scott0o0** [comment]
> hi，我理解你需要将loss_add的逻辑用paddle的接口写，这样在自动求导的时候才能将梯度传过去。否则，paddle是不能将梯度传到numpy数组的。

谢谢解答，但是我在对应的API文档（fluid.layers）中没有找到求tensor变量的转置矩阵的函数。

另外，我在文档中看到的有新增C++OP的方法，但我并没接触过C++。请问在动态图中有没有**简单的方法**可以定义一个新的求loss的接口呢？还是说，只能使用文档中**现有的API**对tensor变量进行编辑才能将梯度传过去呢？

是不是需要等到2.0版本才能在动态图中简单地自定义loss函数，1.8版本的只支持静态图自定义loss函数（包括自定义前向计算逻辑和梯度计算逻辑的这种）。如果可以的话，能否帮我写一个简单的示例呢？

**Comment by zhiqiu** [comment]
> > hi，我理解你需要将loss_add的逻辑用paddle的接口写，这样在自动求导的时候才能将梯度传过去。否则，paddle是不能将梯度传到numpy数组的。
> 
> 谢谢解答，但是我在对应的API文档（fluid.layers）中没有找到求tensor变量的转置矩阵的函数。
> 
> 另外，我在文档中看到的有新增C++OP的方法，但我并没接触过C++。请问在动态图中有没有**简单的方法**可以定义一个新的求loss的接口呢？还是说，只能使用文档中**现有的API**对tensor变量进行编辑才能将梯度传过去呢？
> 
> 是不是需要等到2.0版本才能在动态图中简单地自定义loss函数，1.8版本的只支持静态图自定义loss函数（包括自定义前向计算逻辑和梯度计算逻辑的这种）。如果可以的话，能否帮我写一个简单的示例呢？

转置可以使用fluid.layers.transpose，关于自定义op，如果你对性能没有要求，可以看下自定义python op, [link](https://www.paddlepaddle.org.cn/documentation/docs/zh/advanced_guide/addon_development/new_op/new_python_op.html)。
